### PR TITLE
refactor: make the cmd hold the application instance

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Make the cmd hold the application instance

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#1158

### CODE REVIEW

File: src/cmd/src/bin/greptime.rs [hunk 1]
```diff
@@ -104,14 +149,19 @@ async fn main() -> Result<()> {
...
     tokio::select! {
-        result = cmd.run() => {
+        result = app.run() => {
             if let Err(err) = result {
                 error!(err; "Fatal error occurs!");
             }
         }
         _ = tokio::signal::ctrl_c() => {
             info!("Goodbye!");
+            if let Err(err) = app.stop().await{
```
Comment (MichaelScofield) [inline src/cmd/src/bin/greptime.rs:162]: better place this "if" before line 161, the "Goodbye" log

File: src/meta-srv/src/bootstrap.rs [hunk 2]
```diff
@@ -72,7 +99,7 @@ pub fn router(meta_srv: MetaSrv) -> Router {
         .add_service(admin::make_admin_service(meta_srv))
 }
 
-pub async fn make_meta_srv(opts: MetaSrvOptions) -> Result<MetaSrv> {
+pub async fn build_meta_srv(opts: &MetaSrvOptions) -> Result<MetaSrv> {
```
Comment (MichaelScofield) [inline src/meta-srv/src/bootstrap.rs:102]: it seems `opts` can still be moved here?

File: src/cmd/src/frontend.rs [hunk 3]
```diff
@@ -34,15 +34,33 @@ use snafu::ResultExt;
...
+impl Instance {
+    pub async fn run(&mut self) -> Result<()> {
+        self.frontend
+            .start()
+            .await
+            .context(error::StartFrontendSnafu)
+    }
+
+    pub async fn stop(&self) -> Result<()> {
+        //TODO: handle frontend shutdown
```
Comment (fengjiachun) [inline src/cmd/src/frontend.rs:50]:
```suggestion
        // TODO: handle frontend shutdown
```

File: src/cmd/src/metasrv.rs [hunk 4]
```diff
@@ -14,22 +14,41 @@
...
+    pub async fn run(&mut self) -> Result<()> {
+        self.instance
+            .start()
+            .await
+            .context(error::StartMetaServerSnafu)?;
+        Ok(())
+    }
+
+    pub async fn stop(&self) -> Result<()> {
+        //TODO: handle metasrv shutdown
```
Comment (fengjiachun) [inline src/cmd/src/metasrv.rs:38]:
```suggestion
        // TODO: handle metasrv shutdown
```

File: src/meta-srv/src/bootstrap.rs [hunk 5]
```diff
@@ -39,18 +39,45 @@ use crate::service::store::kv::ResettableKvStoreRef;
...
+
+    pub async fn start(&self) -> Result<()> {
+        self.meta_srv.start().await;
+        bootstrap_meta_srv_with_router(&self.opts.bind_addr, router(self.meta_srv.clone())).await?;
+
+        Ok(())
+    }
+
+    pub async fn close(&self) -> Result<()> {
+        //TODO: shutdown the router
```
Comment (fengjiachun) [inline src/meta-srv/src/bootstrap.rs:64]:
```suggestion
        // TODO: shutdown the router
```

File: src/cmd/src/standalone.rs [hunk 6]
```diff
@@ -133,6 +133,30 @@ impl StandaloneOptions {
...
+            .await
+            .context(StartDatanodeSnafu)?;
+        info!("Datanode instance started");
+
+        self.frontend.start().await.context(StartFrontendSnafu)?;
+        Ok(())
+    }
+
+    pub async fn stop(&self) -> Result<()> {
+        //TODO: handle standalone shutdown
```
Comment (fengjiachun) [inline src/cmd/src/standalone.rs:155]:
```suggestion
        // TODO: handle standalone shutdown
```

**Comment by fengjiachun** [comment]
About `StartCommand`:

Previously, we used `StartCommand` to start `Datanode`, `Metasrv`, `Standalone`, etc. Now its responsibility has changed to just building. Should it still keep its name(StartCommand) unchanged?

@WenyXu @MichaelScofield

**Comment by MichaelScofield** [comment]
I'm ok with `StartCommand`, doesn't see the necessity to rename it.

File: src/meta-srv/src/bootstrap.rs (hunk 2 above)
Comment (waynexia) [reply src/meta-srv/src/bootstrap.rs:102]: I see `opts` is used in many places below, thus we need not (cannot?) remove it I think?

**Review by waynexia — APPROVED** [review]
Looking good to me. PTAL @fengjiachun @MichaelScofield

**Review by fengjiachun — APPROVED** [review]
LGTM

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**